### PR TITLE
WIP: add support for element lifecycle hooks

### PR DESCRIPTION
**Pull request by llllllllll**
It is currently not possible to define a dtype for objects which require some
non-trivial initialization or destruction. This commit adds support for two new
optional functions in `PyArray_ArrFuncs`: `construct` and `destruct`.

If provided, the `construct` function will be called on all new allocations. The
`construct` function should initialize the array with valid elements of the
given type. The `construct` function is allowed to fail, in which case the array
constructor will throw a Python exception.

If provided, the `destruct` function will be called when the buffer is being
deallocated. The `destruct` function should tear down all of the elements in the
array. The destruct function cannot fail.

One explicit goal of this change is to make it easier to define dtypes for C++ types
which are not trivially constructible or trivially destructible. Trivially
constructible means that `memcpy` is a sufficient implementation for copy.
Trivially destructible means that freeing the storage tears down the object
completely. Things like pointers to heap allocated strings would not fall into
this category because they require a deep copy and destruction should free the
data being pointed to.

In theory, the object dtype could be rewritten to use `destruct` to call decref
on each of the elements instead of relying on a special case; however, I am not
sure it is worth doing that yet.

This change does not introduce a new flag for this feature set because
`PyArray_Descr.flags` is a char and all 8 bits are currently being used.
Changing this to `int` may be an ABI incompatible change so I chose to avoid
that. Instead, the function pointers are checked directly before use.

NOTE: This PR does not include tests nor documentation. I would like to discuss this feature with the numpy developers before making too many changes. I have a [proof of concept repo](https://github.com/llllllllll/sbo-string-dtype) which shows how one may use this feature. The linked repo builds and runs with this change applied.

### CODE REVIEW

**Comment by seberg** [comment]
@llllllllll we cannot currently expand that without breaking the ABI and API, e.g. you would have to at least add a new functions to somehow know that this is some "new style" dtype with the additional fields.
Of course we want to change that, but that requires getting it right, and that needs much more ground work and currently that is not going very quickly (because we really should revamp everything a bit more). E.g. see also [NEP 41](https://numpy.org/neps/nep-0041-improved-dtype-support.html) which you are welcome to comment on.

I am also currently not quite convinced that we should go there at all. I heard about your question, and the question is whether it is not sufficient if we provide a typed object array. I.e. you wrap your data in a Python Object, and NumPy knows that all elements are of that object. There are only two minor downsides:

1. You "waste" about 2 pointers, that is part of having memory management.
2. You require the GIL (also part of having memory management really)

Even if the data would be "owned" by the array, rather than the objects, you have such memory management/mutability issues requiring the GIL. What if someone assigns to the array while you are running a ufunc on it.

As a side note, you may want to check out pygeos, who does similar things to what you were doing (with geometrical objects rather than strings).

---

The main downside of just using object arrays right now, is that they are not strictly typed, and any more low level ufuncs would have to have their own full ufunc name, since you cannot register a loop for `object`.

**Comment by seberg** [comment]
@llllllllll I am happy to think about prioritizing this a bit more, but unfortunately I do not think that this is a viable path forward right now.

I realize that the marathon is not very appealing, unfortunately I do not have a clear idea how a sprint can actually get us to a reasonable state myself right now (except through trying to push for a somewhat dirty-preliminary mixed new/old API, which may only work for an intermediate period, even then it would not be super fast).

There is another small thing to consider: If you have decref/incref, we also have to be very careful when copying (including buffers), I am not sure how well that currently works for user-dtypes. I expect this would still be buggy with this fix. Further, we may have to think about `VISIT` in the long run (cyclic reference counting, since I hope we will finally support that at some point).

To be clear, I would love input/support with pushing new things forward to make all of this easier/possible. We are in the scary place that to really allow good user dtypes, we need to revamp everything in my opinion, when at the same time, there seems to be a start for more interest in creating such dtypes based on an old API that is not well vetted :/.

Long story short: I am tempted to close this unless you want to persue it. **But** i would be delighted for any input on a new API (new NEPs) or even direct contribution. That may be hard right now, but I think we can find good ways soon enough, e.g. by contributing test cases, or review time.

**Comment by llllllllll** [comment]
Hey, sorry for not responding right away. Regarding using a PyObject wrapper, I believe I may have answered on a different forum but want to post here to be easier to search:

For the case I was trying to optimize, the overhead of the Python objects dominated the actual data being stored. I wanted to store variable-length strings with no size limit but are otherwise normally under 10 characters. The total overhead for using a PyObject in the array in this case is:
- 'ob_type :: PyTypeObject* = 8 bytes`
- `ob_refcnt :: Py_ssize_t = 8 bytes`
- `<element pointer> :: PyObject* = 8 bytes`
The last pointer is what is stored in the ndarray's data, but it does still take up memory. For small objects, this can be more storage than the data itself. The other issue is that elements in the array are not allocated next to each other. Now, any loop over the array needs to load these 8 byte pointers and then do a dereference to an arbitrary location. Iterating over the raw bytes would be considerably faster for any loop that wanted to actually process the data.

I see why the approach proposed doesn't work with the current ABI. It is unfortunate that the dtype flags are stored in a char and all the bits are used or new bits could be used to test for features like this. Regarding the copy not working: I implemented `copyswap` and `copyswapn` for the type to use the proper copy constructor. Is the issue that this isn't used everywhere it may be needed?

I am okay with closing this PR since a lot more work will be needed and I don't think this is the right way to advocate for the changes I would like. I would be happy to bring my comments to other NEPs discussing new dtype interface ideas.

**Comment by seberg** [comment]
@llllllllll, yeah, I am not 100% sure those are relied on in all places, however, most like this (at least mostly) works fine. Its just hard to be sure, since I doubt anyone every tried, at least not thoroughly...
